pysph/sph/wc/crksph.py

```diff
@@ -373,3 +384,468 @@ def loop(self, d_idx, s_idx, d_ai, d_gradai, d_cwij, d_bi, d_gradbi, s_ai,
         if change < self.tol:
             for i in range(d):
                 DWIJ[i] = res[i]
+
+
+class NumberDensity(Equation):
+    r"""**Number Density**
+
+    From [CRKSPH2017], equation (75):
+
+    .. math::
+        V_{i}^{-1} = \sum_{j} W_{i}
+    """
+    def initialize(self, d_idx, d_V):
+        d_V[d_idx] = 0.0
+
+    def loop(self, d_idx, d_V, WIJ):
+        d_V[d_idx] += WIJ
+
+
+class SummationDensityCRKSPH(Equation):
+    r"""**Summation Density CRKSPH**
+
+    From [CRKSPH2017], equation (76):
+
+    .. math::
+        \rho_{i} = \frac{\sum_j m_{ij} V_j W_{ij}^R}
+        {\sum_{j} V_{j}^{2} W_{ij}^R}
+    """
+
+    def initialize(self, d_idx, d_rho, d_rhofac):
+        d_rho[d_idx] = 0.0
+        d_rhofac[d_idx] = 0.0
+
+    def loop(self, d_idx, s_idx, d_m, d_rho, d_rhofac, s_V, WIJ, d_cwij):
+        Vj = 1.0/s_V[s_idx]
+        fac = Vj * d_cwij[d_idx] * WIJ
+        d_rho[d_idx] += d_m[d_idx] * fac
```

Comment:
I figured since we are not changing the mass anywhere s_m[s_idx] will be equal to d_m[d_idx] for similar material, and just use d_m[d_idx] as it works for both similar and different material.